File: ReadNote/douban.py
# @Author: Lumos

from datetime import date
import requests
from bs4 import BeautifulSoup
import re

# 模板文件路径
TEMPLATEPATH = r"F:/OneDrive/MyPostgraduate/Template/ReadNote.md"

# 生成的笔记所在的目录
SAVEDIR = r"F:/OneDrive/MyPostgraduate/3日常/读书/"

# requests所需的代理参数，不知道为什么我不设置的话会报错
PROXIES = {
    'http': 'http://127.0.0.1:7890',
    'https': 'http://127.0.0.1:7890'
}

# 获取基础信息（标题以及详细信息的网址）
def getSimpleInfo(isbn:str) -> dict :
    url = f"https://m.douban.com/search/?query={isbn}"
    headers = {
    'sec-ch-ua': '" Not A;Brand";v="99", "Chromium";v="98", "Google Chrome";v="98"',
    'sec-ch-ua-mobile': '?0',
    'sec-ch-ua-platform': '"Windows"',
    'Upgrade-Insecure-Requests': '1',
    'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/98.0.4758.102 Safari/537.36',
    'Referer': 'https://m.douban.com/search/?query=9787210105893',
    }

    response = requests.get(url,headers=headers,proxies=PROXIES)
    soup = BeautifulSoup(response.content,"lxml")

    temp = soup.select("div[class=subject-info] > span")
    if temp:
        title = temp[0].text
        detailUrl = "https://m.douban.com"+soup.select("ul>li>a")[0].get("href")
    else:
        return None
    
    return {"title":title,"detailUrl":detailUrl}


# 获取详细信息
def getDetailInfo(detailUrl:str) -> dict:
    headers = {
    'Connection': 'keep-alive',
    'Upgrade-Insecure-Requests': '1',
    'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/98.0.4758.102 Safari/537.36',
    'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9',
    'sec-ch-ua': '" Not A;Brand";v="99", "Chromium";v="98", "Google Chrome";v="98"',
    'sec-ch-ua-mobile': '?0',
    'sec-ch-ua-platform': '"Windows"',
    'Sec-Fetch-Site': 'same-site',
    'Sec-Fetch-Mode': 'navigate',
    'Sec-Fetch-User': '?1',
    'Sec-Fetch-Dest': 'document',
    'Referer': 'https://m.douban.com/',
    'Accept-Language': 'en-US,en;q=0.9,zh-CN;q=0.8,zh;q=0.7',
    }

    response = requests.get(detailUrl,headers=headers,proxies=PROXIES)
    soup = BeautifulSoup(response.content,"lxml")
    
    ## 书名
    name = soup.find("h1").text.replace("\n","")
    title = "\""+name+"\""

    ## 封面图片
    coverUrl = soup.select("div[id=mainpic]>a")[0].get("href")

    ## 作者和译者、原作名（如果有译者的话）、页数
    spanList=soup.find(id="info").select("span")
    author=""
    transAuthor=""
    originalName = title
    pages = ""
    for i in spanList:
        if "作者" in i.text and i.find_next_sibling("a")!=None:
            author = re.sub(r"[\s\[]","",i.find_next_sibling("a").text).replace("]"," ")
        elif "译者" in i.text and i.find_next_sibling("a")!=None:
            transAuthor = re.sub(r"[\s\[]","",i.find_next_sibling("a").text).replace("]"," ")
        elif "原作名" in i.text:
            originalName = "\""+i.next_sibling.text.strip()+"\""
        elif "页数" in i.text:
            pages = i.next_sibling.text.strip()

    ## ISBN号码，这个字段没有在标签下所以不太好定位，但是isbn一般为13个数字，所以直接用正则匹配
    patt = re.compile(r"\d{10,13}")
    isbn = patt.findall(soup.select("div[id=info]")[0].text)[0]

    ## 豆瓣评分
    rating = soup.select("div[id=interest_sectl]>div>div>strong")[0].text.replace(" ","")

    ## 简介
    # introList = soup.select("div[class=intro]")
    # ### 书籍简介
    # intro = ''
    # for i in introList[0].select("p"):
    #     intro = intro+i.text+"\n"
        
    # # for i in soup.select("div[id=link-report] > div> div>p"):
    # #     intro = intro+i.text+"\n"

    # ### 作者简介
    # authorIntro = ''
    # for i in introList[-1].select('p'):
    #     authorIntro = authorIntro+i.text+"\n"
    ### 书籍简介
    intro = ''
    for i in soup.select("h2")[0].find_next_sibling("div").select("div[class=intro]")[-1].select("p"):
        intro = intro+i.text+"\n"
    
    ### 作者简介
    authorIntro = ''
    for i in soup.select("h2")[1].find_next_sibling("div").select("div[class=intro]")[-1].select("p"):
        authorIntro = authorIntro+i.text+"\n"
    
    ## 部分原文摘录（如果有的话）
    pattern = re.compile(r"[\S ]+")
    quoteList = soup.select("figure")
    sourceList = soup.select("figcaption")
    if quoteList:
        quote1 = pattern.findall(quoteList[0].contents[0])[0].replace("(","").strip()+"\n> "+sourceList[0].text.replace("\n","")
        quote2 = pattern.findall(quoteList[1].contents[0])[0].replace("(","").strip()+"\n> "+sourceList[1].text.replace("\n","")
    else:
        quote1 = quote2 = None
    
    ## 标签
    tagPattern = re.compile(r"(?<=:)[\u4e00-\u9fa5·]+")
    tags = tagPattern.findall(soup.select("script")[-3].text)
    tags.append("book")
    
    ## 相关书籍
    relatedBooks = [i.text.replace("\n","").replace(" ","") for i in soup.select("div[id=db-rec-section]>div>dl>dd")]

    ## 当前日期，相当于文件创建日期
    today = date.today().strftime("%Y-%m-%d")

    detailInfo = {
        "title":title,
        "name":name,
        "coverUrl":coverUrl,
        "author":author,
        "transAuthor":transAuthor,
        "originalName":originalName,
        "rating":rating,
        "pages":pages,
        "tags":tags,
        "relatedBooks":relatedBooks,
        "link":detailUrl,
        "isbn":isbn,
        "today":today,
        "intro":intro,
        "authorIntro":authorIntro,
        "quote1":quote1,
        "quote2":quote2
    }
    return detailInfo


# 生成阅读笔记
def createNote(detailInfo:dict)-> None:
    with open(TEMPLATEPATH,"r",encoding="utf-8") as temp:
        template = temp.read()
    
    content = template.format(**detailInfo)
    
    with open(SAVEDIR + detailInfo["name"] + ".md","w",encoding="utf-8") as f:
        f.write(content)

if __name__ == "__main__":
    isbn = "9787530211267" #《一个陌生女人的来信》
    url = getSimpleInfo(isbn)["detailUrl"]
    from time import sleep
    sleep(2)
    createNote(getDetailInfo(url))

File: CreateReadNote/QuickAdd版本/douban.js
// Author: @Lumos
// Url: https://github.com/LumosLovegood
async function douban(QuickAdd){
    const isbn = await QuickAdd.quickAddApi.inputPrompt(
        "请输入书籍背后的13位ISBN码："
    );
    if(isbn.length!=13){
        new Notice("ISBN码位数错误");
        throw new Error("ISBN码位数错误");
    }
    let simpleInfo =await getBookUrl(isbn);
    if(!simpleInfo){
        new Notice("无法识别此ISBN码");
        throw new Error("无法识别此ISBN码");
    }
    let url = simpleInfo.url;
    new Notice("准备获取《"+simpleInfo.title+"》的内容信息",1000)
    let bookInfo = await getDetailInfo(url)
    if(!bookInfo){
        new Notice("获取内容失败");
        throw new Error("获取内容失败");
    }
    new Notice("笔记已生成！",500);
    // 获取今日日期
    const date = window.moment().format("gggg-MM-DD")
    bookInfo.today = date;

    QuickAdd.variables = {
        ...bookInfo
    };
}

async function getBookUrl(isbn){
    url = "https://m.douban.com/search/?query="+isbn;
    let searchUrl = new URL(url);
    const res = await request({
      url: searchUrl.href,
      method: "GET",
      cache: "no-cache",
      headers: {
        "Content-Type": "text/html; charset=utf-8",
        'sec-ch-ua': '" Not A;Brand";v="99", "Chromium";v="98", "Google Chrome";v="98"',
        'sec-ch-ua-mobile': '?0',
        'sec-ch-ua-platform': '"Windows"',
        'Upgrade-Insecure-Requests': '1',
        'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/98.0.4758.102 Safari/537.36'
      }
    });

    if(!res){
        return null;
    }

    let p = new DOMParser();
    let doc = p.parseFromString(res, "text/html");
    let title = doc.querySelector("div.subject-info span").textContent;
    let detailUrl = String(doc.querySelector("ul li a").href).replace("app://obsidian.md","https://m.douban.com");
    if (!detailUrl){
        return null;
    }
    let simpleInfo={};
    simpleInfo.title=title;
    simpleInfo.url = detailUrl;
    return simpleInfo;
}

async function getDetailInfo(url){
    let bookUrl = new URL(url);
    const res = await request({
        url: bookUrl.href,
        method: "GET",
        cache: "no-cache",
        headers: headers
    });
    let p = new DOMParser();
    let doc = p.parseFromString(res, "text/html");
    let $ = s => doc.querySelector(s);
    let $2 = z => doc.querySelectorAll(z);


    let bookInfo = {}; 
    //书名、作者、ISBN、封面
    let name = $("meta[property='og:title']")?.content;
    let title = "\""+name+"\""; //用于放到front matter里，加引号避免因为包含特殊字符导致ymal解析错误
    let author = $("meta[property='book:author']")?.content.replace(/[\[\]\(\)（）]/g,"");
    let isbn = $("meta[property='book:isbn']")?.content;
    let cover = $("meta[property='og:image']")?.content;
    
    //其他信息(译者、原作名、页数)
    let text = $("#info")?.textContent.replace("\n","");
    let transAuthor = text.match(/(?<=译者:\s*)\S+\s?\S+/g)?text.match(/(?<=译者:\s*)\S+\s?\S+/g)[0].trim():"";
    let originalName = text.match(/(?<=原作名:\s*)[\S ]+/g)?text.match(/(?<=原作名:\s*)[\S ]+/g)[0].trim():"";
    let pages = text.match(/(?<=页数:\s*)[\S ]+/g)?text.match(/(?<=页数:\s*)[\S ]+/g)[0].trim():"";
    let publisher = text.match(/(?<=出版社:\s*)\S+\s?\S+/g)?text.match(/(?<=出版社:\s*)\S+\s?\S+/g)[0].trim():"";

    //豆瓣评分
    let rating = $("div#interest_sectl div div strong")?.textContent.replace(/\s/g,"");

    //书籍和作者简介，这一块儿不同类型的书对应的网页结构都不太一样，尽力做兼容了，还有问题我也没办法 \摊手
    let intro = "";
    let authorIntro = "";
    var temp1 = $("h2");
    if(temp1.innerText.includes("内容简介")){
        var temp2 = temp1.nextElementSibling.querySelectorAll("div.intro")
        var temp3 = temp2[temp2.length-1].querySelectorAll("p");
        for(var i=0;i<temp3.length;i++){
            intro = intro+temp3[i].textContent+"\n";
        }
        try{
            temp2 = $2("h2")[1].nextElementSibling.querySelectorAll("div.intro");
            temp3 = temp2[temp2.length-1].querySelectorAll("p");
            for(var i=0;i<temp3.length;i++){
                authorIntro = authorIntro+temp3[i].textContent+"\n";
            }
        }catch(e){
            new Notice("没有作者简介");
        }        
    }else if(temp1.innerText.includes("作者简介")){
        var temp2 = temp1.nextElementSibling.querySelectorAll("div.intro")
        var temp3 = temp2[temp2.length-1].querySelectorAll("p");
        for(var i=0;i<temp3.length;i++){
            authorIntro = authorIntro+temp3[i].textContent+"\n";
        }
    }

    //原文摘录
    let quote1 = "";
    let quote2 = "";
    let quoteList = $2("figure");
    let sourceList = $2("figcaption");
    if(quoteList){
        quote1 = quoteList[0].childNodes[0].textContent.replace(/\(/g,"").trim()+"\n"+sourceList[0].textContent.replace(/\s/g,"");
        quote2 = quoteList[1].childNodes[0].textContent.replace(/\(/g,"").trim()+"\n"+sourceList[1].textContent.replace(/\s/g,"");
    }

    //豆瓣常用标签，记得之前这一块儿网页元素里是有的，后来找不到了，但是尝试性源代码全文搜索的时候 在Script标签里找到了，但是感觉随时会改。
    var temp = $2("script");
    let tags = temp[temp.length-3].textContent.match(/(?<=:)[\u4e00-\u9fa5·]+/g);
    tags.push("book");

    //相关书籍，仿佛这个信息也没啥用，但是能加就加了
    let relatedBooks = [];
    temp = $2("div#db-rec-section div dl dd");
    if(temp){
        for(var i=0;i<temp.length;i++){
            relatedBooks.push(temp[i].textContent.replace(/\s/g,""));
        }
    }

    bookInfo.name = name;
    bookInfo.title=title;
    bookInfo.author=author;
    bookInfo.transAuthor=transAuthor;
    bookInfo.coverUrl=cover;
    bookInfo.originalName=originalName;
    bookInfo.pages=pages;
    bookInfo.publisher=publisher;
    bookInfo.intro=intro;
    bookInfo.isbn=isbn;
    bookInfo.rating=rating;
    bookInfo.authorIntro =authorIntro;
    bookInfo.quote1=quote1;
    bookInfo.quote2=quote2;
    bookInfo.tags=tags;
    bookInfo.relatedBooks=relatedBooks;
    bookInfo.link = url;

    // 如果为空的话，quickadd会出现提示框让自己填，太麻烦了，所以先填一个默认空值
    for(var i in bookInfo){
        if(bookInfo[i]==""){
            bookInfo[i]="Not Found.";
        }
    }

    return bookInfo;
}
module.exports =  douban
